test/common.py

```diff
@@ -449,6 +449,44 @@ def crossplatform(f):
   return f
 
 
+# Without EMTEST_ALL_ENGINES set we only run tests in a single VM by default.
+# However, for some tests we know that cross-VM differences may happen and
+# so are worth testing, and they should be marked with this decorator which creates
+# N different variants of the test, one for each engine.
+def all_engines(f):
+  assert callable(f)
+
+  if len(config.JS_ENGINES) == 1 or EMTEST_ALL_ENGINES:
+    return f
+
+  @wraps(f)
+  def metafunc(self, engine, *args, **kwargs):
+    self.set_setting('ENVIRONMENT', 'web,node,shell')
+    self.js_engines = [engine]
+    f(self, *args, **kwargs)
+
+  engine_mapping = {}
+  for engine in config.JS_ENGINES:
+    if not engine_mapping:
+      # For the first engine we use no suffix at all.
+      # This uffnsures that one of the tests matches the original name, meaning that
+      # you can still run just `test_foo` and always get the first engine without
+      # needing to write, for example, `test_foo_node`.
+      name = ''
+    else:
+      basename = os.path.basename(engine[0])
+      name = basename
+      suffix = 1
+      while name in engine_mapping:
+        name = f'{basename}_{suffix}'
+        suffix += 1
+    engine_mapping[name] = (engine,)
+
+  parameterize(metafunc, engine_mapping)
```

Comment:
Does this not run the entire test in each engine, while before we built once and just executed the JS once per engine?

Reply:
Wait, none of our CI setups test multiple engines? Don't we set up V8 on circleCI?

Reply:
We have CI tests that configure different engines, but never more than one engine at the same time.

`@all_engines` is a no-op unless you have multiple engines configured at the same time.

Reply:
Actually this might not be totally true, I think for v8 we *do* configure it to be in addition to node (i.e. we append it to JS_ENGINES) but for jsc and spidermonkey testing we always just do `JS_ENGINES = [spidermonkey]` or `JS_ENGINES = [jsc]`.